controllers/raycluster_controller.go

```diff
@@ -0,0 +1,424 @@
+func (r *RayClusterReconciler) reconcileAutoscaler(ctx context.Context, rc *dcv1alpha1.RayCluster) error {
+	if rc.Spec.Autoscaling == nil {
+		hpa := &autoscalingv2beta2.HorizontalPodAutoscaler{
+			ObjectMeta: ray.HorizontalPodAutoscalerObjectMeta(rc),
+		}
+		return r.deleteIfExists(ctx, hpa)
+	}
+
+	hpa, err := ray.NewHorizontalPodAutoscaler(rc)
+	if err != nil {
+		return err
+	}
+	if err = r.createOrUpdateOwnedResource(ctx, rc, hpa); err != nil {
+		return fmt.Errorf("failed to reconcile horizontal pod autoscaler: %w", err)
+	}
+
+	return nil
+}
+
+// reconcileDeployments creates separate Ray head and worker deployments that
+// will collectively comprise the execution agents of the cluster.
+func (r *RayClusterReconciler) reconcileDeployments(ctx context.Context, rc *dcv1alpha1.RayCluster) error {
+	head, err := ray.NewDeployment(rc, ray.ComponentHead)
+	if err != nil {
+		return err
+	}
+	if err = r.createOrUpdateOwnedResource(ctx, rc, head); err != nil {
+		return fmt.Errorf("failed to create head deployment: %w", err)
+	}
+
+	worker, err := ray.NewDeployment(rc, ray.ComponentWorker)
+	if err != nil {
+		return err
+	}
+	if err = r.createOrUpdateOwnedResource(ctx, rc, worker); err != nil {
+		return fmt.Errorf("failed to create worker deployment: %w", err)
+	}
+
+	return nil
+}
+
+// createOrUpdateOwnedResource should be used to manage the lifecycle of namespace-scoped objects.
+//
+// The CR will become the "owner" of the "controlled" object and cleanup will
+// occur automatically when the CR is deleted.
+//
+// The controller resource will be created if it's missing.
+// The controller resource will be updated if any changes are applicable.
+// Any unexpected api errors will be reported.
+func (r *RayClusterReconciler) createOrUpdateOwnedResource(ctx context.Context, owner metav1.Object, controlled client.Object) error {
+	if err := ctrl.SetControllerReference(owner, controlled, r.Scheme); err != nil {
+		return err
+	}
+
+	log := r.getLogger(ctx)
+	found := controlled.DeepCopyObject().(client.Object)
+	err := r.Get(ctx, client.ObjectKeyFromObject(controlled), found)
+
+	if apierrors.IsNotFound(err) {
+		if err = PatchAnnotator.SetLastAppliedAnnotation(controlled); err != nil {
+			return err
+		}
+
+		log.Info("creating controlled object", "object", controlled)
+		return r.Create(ctx, controlled)
+	}
+	if err != nil {
+		return err
+	}
+
+	patchResult, err := PatchMaker.Calculate(found, controlled, patch.IgnoreStatusFields())
+	if err != nil {
+		return err
+	}
+	if patchResult.IsEmpty() {
+		return nil
+	}
+
+	log.V(1).Info("applying patch to object", "object", controlled, "patch", string(patchResult.Patch))
+	if err = PatchAnnotator.SetLastAppliedAnnotation(controlled); err != nil {
+		return err
+	}
+
+	controlled.SetResourceVersion(found.GetResourceVersion())
+	if modified, ok := controlled.(*corev1.Service); ok {
+		current := found.(*corev1.Service)
+		modified.Spec.ClusterIP = current.Spec.ClusterIP
+	}
```

Comment:
this ... seems a little odd and scary (i.e. feels like we should either have immutable services or somehow copy all the stateful-ish fields)